src/gcp_airflow_foundations/operators/airflow/external_task.py

```diff
@@ -0,0 +1,157 @@
+            raise AirflowException(
+                f"Duplicate values provided as allowed "
+                f"`{self.allowed_states}` and failed states `{self.failed_states}`"
+            )
+       
+        self.execution_delta = execution_delta
+        self.external_dag_ids = self.get_external_dag_ids()
+
+    @provide_session
+    def poke(self, context, session=None):
+        if self.execution_delta:
+            dttm = context['execution_date'] - self.execution_delta
+        else:
+            dttm = context['execution_date']
+
+        dttm_filter = dttm if isinstance(dttm, list) else [dttm]
+        serialized_dttm_filter = ','.join(dt.isoformat() for dt in dttm_filter)
+
+        count_allowed = self.get_count(dttm_filter, session, self.allowed_states)
+
+        count_failed = -1
+        if self.failed_states:
+            count_failed = self.get_count(dttm_filter, session, self.failed_states)
+
+        if count_failed == len(dttm_filter):
+            pass
+
+        return count_allowed == len(dttm_filter)
+
+    def get_count(self, dttm_filter, session, states) -> int:
+        """
+        Get the count of records against dttm filter and states
+        :param dttm_filter: date time filter for execution date
+        :type dttm_filter: list
+        :param session: airflow session object
+        :type session: SASession
+        :param states: task or dag states
+        :type states: list
+        :return: count of record against the filters
+        """
+        TI = TaskInstance
+        DR = DagRun
+        
+        expected_count = len(self.external_dag_ids)
+        count = (
+            session.query(func.count())
+            .filter(
+                DR.dag_id.in_(self.external_dag_ids),
+                DR.state.in_(states),
+                DR.execution_date.in_(dttm_filter),
+            )
+            .scalar()
+        )
+
+        self.log.info(
+            'Current count of completed tasks is %s. The expected DAG count is %s',
+            count,
+            expected_count,
+        )
+            
+        return count / expected_count
+
+    @provide_session
+    def get_external_dag_ids(self, context, session=None) -> list:
+        """
+        Retrieve a list of external DAG IDs based on the provided source & table combinations
+        """
+        schedule_interval = context['dag'].schedule_interval
+
+        external_dag_ids = []
+        sources = [i for i in self.external_source_tables]
+
+        query = session.query(DagModel).filter(DagModel.is_active==True).all()
+        active_dags = [i.dag_id for i in query]
+        schedule_map = {i.dag_id:i.schedule_interval for i in query}
+
+        if not active_dags:
+            raise AirflowException(f'No active dags found')
+
+        for source in sources:
+            tables = self.external_source_tables[source]
+            source_dags = [i for i in active_dags if source in i]
```

Comment:
maybe I am over optimizing, but it would be more efficient if active dags was a map of source -> list of dags. This way lookup is O(1) instead of O(n)

Reply:
Using the underscore as delimiter in the DAG naming we can readily extrapolate the source name from the DAG ID, which enables us to create a map for the lookup.